pkg/reconciler/eventtype.go

```diff
@@ -0,0 +1,193 @@
+/*
+Copyright 2019 The Knative Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reconciler
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/util/validation"
+
+	eventingv1alpha1 "github.com/knative/eventing/pkg/apis/eventing/v1alpha1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+// Only allow alphanumeric, '-' or '.'.
+var validChars = regexp.MustCompile(`[^-\.a-z0-9]+`)
+
+// Reconciler is a helper struct that can be used by any source in order to reconcile its EventTypes.
+type EventTypeReconciler struct {
+	Client client.Client
+	Scheme *runtime.Scheme
+}
+
+// ReconcilerArgs are the arguments needed to reconcile EventTypes.
+type EventTypeReconcilerArgs struct {
+	EventTypeSpecs []eventingv1alpha1.EventTypeSpec
+	Namespace      string
+	Labels         map[string]string
+}
+
+// syncEventTypes is a helper struct used to sync EventTypes during the reconciliation process.
+type syncEventTypes struct {
+	ToCreate []eventingv1alpha1.EventType
+	ToDelete []eventingv1alpha1.EventType
+}
+
+// ReconcileEventTypes reconciles the EventTypes taken from 'args', and sets 'owner' as the controller.
+func (r *EventTypeReconciler) ReconcileEventTypes(ctx context.Context, owner metav1.Object, args *EventTypeReconcilerArgs) error {
+	current, err := r.getEventTypes(ctx, args.Namespace, args.Labels, owner)
+	if err != nil {
+		return err
+	}
+
+	expected, err := r.makeEventTypes(args, owner)
+	if err != nil {
+		return err
+	}
+	sync := r.getEventTypesToSync(current, expected)
+	for _, eventType := range sync.ToCreate {
+		err = r.Client.Create(ctx, &eventType)
+		if err != nil {
+			return err
+		}
+	}
+	// Deletion might not be needed, but just to make sure we dedupe EventTypes
+	// in case some source controller has a bug.
+	for _, eventType := range sync.ToDelete {
+		err = r.Client.Delete(ctx, &eventType)
+		if err != nil && !errors.IsNotFound(err) {
+			return err
+		}
+	}
+	return nil
+}
+
+// getEventTypes returns the EventTypes controlled by 'owner'.
+func (r *EventTypeReconciler) getEventTypes(ctx context.Context, namespace string, lbs map[string]string, owner metav1.Object) ([]eventingv1alpha1.EventType, error) {
+	eventTypes := make([]eventingv1alpha1.EventType, 0)
```

Comment:
I think we normally do `eventTypes := []eventingv1alpha1.EventType{}`

Reply:
(You can ignore this comment given https://github.com/knative/eventing-sources/pull/300#discussion_r280204402)